src/core/project/qgsproject.cpp

```diff
@@ -1305,6 +1308,87 @@ void QgsProject::setAvoidIntersectionsMode( const Qgis::AvoidIntersectionsMode m
   emit avoidIntersectionsModeChanged();
 }
 
+static  QgsMapLayer::ReadFlags projectFlagsToLayerReadFlags( Qgis::ProjectReadFlags projectReadFlags, Qgis::ProjectFlags projectFlags )
+{
+  QgsMapLayer::ReadFlags layerFlags = QgsMapLayer::ReadFlags();
+  if ( projectReadFlags & Qgis::ProjectReadFlag::DontResolveLayers )
+    layerFlags |= QgsMapLayer::FlagDontResolveLayers;
+  // Propagate trust layer metadata flag
+  if ( ( projectFlags & Qgis::ProjectFlag::TrustStoredLayerStatistics ) || ( projectReadFlags & Qgis::ProjectReadFlag::TrustLayerMetadata ) )
+    layerFlags |= QgsMapLayer::FlagTrustLayerMetadata;
+  // Propagate open layers in read-only mode
+  if ( ( projectReadFlags & Qgis::ProjectReadFlag::ForceReadOnlyLayers ) )
+    layerFlags |= QgsMapLayer::FlagForceReadOnly;
+
+  return layerFlags;
+}
+
+void QgsProject::preloadProviders( const QVector<QDomNode> &asynchronusLayerNodes,
+                                   const QgsReadWriteContext &context,
+                                   QMap<QString, QgsDataProvider *> &loadedProviders,
+                                   QgsMapLayer::ReadFlags layerReadFlags,
+                                   int totalProviderCount )
+{
+
+  QVector<QgsRunnableProviderCreator *> runnables;
+  QThreadPool threadPool;
+
+  QgsSettings settings;
+  if ( settings.contains( QStringLiteral( "provider-preloading/max-thread-count" ) ) )
+  {
+    threadPool.setMaxThreadCount( settings.value( QStringLiteral( "provider-preloading/max-thread-count" ) ).toInt() );
+  }
+  else
+  {
+    threadPool.setMaxThreadCount( QThread::idealThreadCount() );
+    settings.setValue( QStringLiteral( "provider-preloading/max-thread-count" ), threadPool.maxThreadCount() );
+  }
+
+  int i = 0;
+  int validLayerCount = 0;
+  QEventLoop *loop = new QEventLoop( this );
+  for ( const QDomNode &node : asynchronusLayerNodes )
+  {
+    const QDomElement layerElement = node.toElement();
+    QString layerId = layerElement.namedItem( QStringLiteral( "id" ) ).toElement().text();
+    QString provider = layerElement.namedItem( QStringLiteral( "provider" ) ).toElement().text();
+    QString dataSource = layerElement.namedItem( QStringLiteral( "datasource" ) ).toElement().text();
+
+    dataSource = QgsProviderRegistry::instance()->relativeToAbsoluteUri( provider, dataSource, context );
+
+    QgsDataProvider::ProviderOptions options( {context.transformContext()} );
+    QgsDataProvider::ReadFlags flags = QgsMapLayer::providerReadFlags( node, layerReadFlags );
+    flags.setFlag( QgsDataProvider::NoResquestCredentials, true );
```

Comment:
is it worth adding a comment that we don't want to request credentials, because this would cause deadlocks if credentials are requested from worker thread (while main thread is waiting for the workers to finish)